providers/implementations/digests/sha3_prov.c

```diff
@@ -664,6 +669,21 @@ static int KECCAK_Deserialize(KECCAK1600_CTX *c, int impl_id,
     if (val != (uint64_t)c->md_size)
         return 0;
 
+    /* check that block_size is congruent with the initialized value */
+    p = OPENSSL_load_u64_le(&val, p);
+    if (val != c->block_size)
+        return 0;
+    /* check that bufsz does not exceed block_size */
+    p = OPENSSL_load_u64_le(&val, p);
+    if (val > c->block_size)
+        return 0;
+    c->bufsz = (size_t)val;
+    p = OPENSSL_load_u64_le(&val, p);
+    if (val != c->pad)
+        return 0;
+    p = OPENSSL_load_u64_le(&val, p);
+    c->xof_state = (int)val;
```

Comment:
Are we allowed to be in certain states here? non xof's would never be in squeeze state. What about final state?

Reply:
@slontis we are not validating the content, only ensuring a random bit flip on a disc will not cause a crash by going out of memory.

To be able to validate states we'd have to introspect each algorithm and know what is permissible there or not, and I really see no point in doing that because applications are not supposed to pass in untrusted data anyway.